File: cpp/boyerMoore.cpp
// efficient for large texts and patterns
// badCharHeuritic preprocesses pattern to create a bad character table
// this table stores last occurrence of each character in pattern 

#include <iostream>
#include <cstring>

using namespace std;

void badCharHeuristic(string str, int size, int badchar[256])
{
  int i;
  for(i = 0; i < 256; i++)
    badchar[i] = -1;
  for(i = 0; i < size; i++)
    badchar[(int)str[i]] = i;
}

void search(string txt, string pat)
{
  int m = pat.length();
  int n = txt.length();

  int badchar[256];
  badCharHeuristic(pat, m, badchar);

  int s = 0; // s is shift of pattern 
  while(s <= (n - m)) {
    int j = m - 1;
    while(j >= 0 && pat[j] == txt[s+j])
      j--;
    if(j < 0) {
      cout << "pattern found at index " << s << end1;
      s += (s + m < n) ? m - badchar[txt[s+m]] : 1;
    }
    else {
      s += max(1, j - badchar[txt[s+j]]);
    }
  }
}
int main()
{
  string txt = "ABAAABCDABAAABAA";
  string pat = "ABAA";
  search(txt, pat);
  return 0;
}

File: typescript/boyerMoore.ts
function boyerMooreSearch(text: string, pattern: string): number[] {
  const results: number[] = [];

  // preprocessing; create bad character shift table
  const badCharShift = new Map<string, number>();
  for(let i = 0; i < pattern.length - 1; i++) {
    badCharShift.set(pattern[i], pattern.length - i - 1);
  }

  // preprocessing; create good suffix shift table
  const goodSuffixShift = new Array(pattern.length).fill(pattern.length);
  let suffix = new Array(pattern.length).fill(false);
  for(let i = pattern.length - 1, j = pattern.length - 1; i >= 0; i--, j--) {
    if(pattern[i] === pattern[j]) {
      suffix[i] = true;
      goodSuffixShift[i] = goodSuffixShift[i + 1];
    } else {
      goodSuffixShift[i] = j - i;
      j = pattern.length - 1;
    }
  }

  // search phase
  let i = 0;
  while(i <= text.length - pattern.length) {
    let j = pattern.length - 1;
    while(j >= 0 && text[i + j] === pattern[j]) {
      j--;
    }
    if(j < 0) {
      results.push(i);
      i += goodSuffixShift[0];
    } else {
      i += Math.max(goodSuffixShift[j + 1], badCharShift.get(text[i + j]) || pattern.length);
    }
  }
  return results;
}
